rosbag2_cpp/src/rosbag2_cpp/rmw_implemented_serialization_format_converter.cpp

```diff
@@ -0,0 +1,132 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "./rmw_implemented_serialization_format_converter.hpp"
+
+#include <memory>
+#include <sstream>
+#include <string>
+
+#include "ament_index_cpp/get_resources.hpp"
+#include "rcpputils/find_library.hpp"
+#include "rcpputils/shared_library.hpp"
+#include "rmw/rmw.h"
+#include "rosbag2_cpp/converter_interfaces/serialization_format_converter.hpp"
+#include "rosbag2_cpp/logging.hpp"
+
+namespace
+{
+/// Get a function with a given signature from a shared library, with error checking.
+template<typename T>
+T get_function_from(const char * function_name, std::shared_ptr<rcpputils::SharedLibrary> library)
+{
+  if (!library->has_symbol(function_name)) {
+    std::stringstream ss;
+    ss << "Converter could not find expected symbol '" << function_name <<
+      "' in rmw implementation library " << library->get_library_path();
+    throw std::runtime_error{ss.str()};
+  }
+  T loaded_function = nullptr;
+  // Function expected to return a value because has_symbol was checked first.
+  loaded_function = (decltype(loaded_function))(library->get_symbol(function_name));
+  return loaded_function;
+}
+}  // namespace
+
+namespace rosbag2_cpp
+{
+
+class RMWImplementedConverterImpl
+{
+public:
+  explicit RMWImplementedConverterImpl(const std::string & format)
+  {
+    const std::string current_implementation_format{rmw_get_serialization_format()};
+    if (current_implementation_format == format) {
+      deserialize_fn_ = &rmw_deserialize;
+      serialize_fn_ = &rmw_serialize;
+      return;
+    }
+
+    auto packages_with_prefixes = ament_index_cpp::get_resources("rmw_typesupport");
+
+    for (const auto & package_prefix_pair : packages_with_prefixes) {
+      ROSBAG2_CPP_LOG_INFO_STREAM("Found package " << package_prefix_pair.first);
+      const auto & pkg = package_prefix_pair.first;
+      if (pkg == "rmw_implementation") {
+        continue;
+      }
+
+      const auto libpath = rcpputils::find_library_path(pkg);
+      if (libpath.empty()) {
+        ROSBAG2_CPP_LOG_ERROR_STREAM(
+          "Unexpectedly could not find library for RMW implementation " << pkg);
+        continue;
+      }
+      library_ = std::make_shared<rcpputils::SharedLibrary>(libpath);
+
+      auto get_format_fn = get_function_from<decltype(&rmw_get_serialization_format)>(
+        "rmw_get_serialization_format", library_);
+      const char * impl_serialization_format = get_format_fn();
+      if (impl_serialization_format == format) {
+        deserialize_fn_ = get_function_from<decltype(deserialize_fn_)>(
+          "rmw_deserialize", library_);
+        serialize_fn_ = get_function_from<decltype(serialize_fn_)>(
+          "rmw_serialize", library_);
```

Comment:
might not be an immediate problem, but I wanted to mention it anyway.

For Fast-RTPS we were certain that the two functions (`rmw_deserialize, rmw_serialize`) could be called without any state context - that is `rmw_init`. However in this generic case here, can we always guarantee that I can call these functions without any prior calls to initialize the RMW?

Reply:
There may not be such a guarantee. `rmw_init` creates an `rmw_context_t`, which you have to pass to certain API functions. For RMW functions that do not require an `rmw_context_t`, it's unclear to me whether `rmw_init` was required to have been called. It may not hurt to call it here and hold onto a context in the converter.